Multiple improvements to Spike smoke-test workflow

JanMatCodasip opened the pull request:
Change-Id: I5b8e9200c5d8cbaa3116bac565e009089bb6b36b

JanMatCodasip committed:
Multiple improvements to Spike smoke-test workflow

These changes have been made:

- use checkout action v3 (not v2) - silences a Github warning
- cache dependencies to speed-up the process (Spike + toolchain)
- reorder actions so that cached items are handled first
- move dependencies to /opt/riscv for easier caching
- archive logs from the tests (downloadable artifact)
- more descriptive names for some steps
- changed 'apt' to 'apt-get' to supress a warning

Change-Id: I5b8e9200c5d8cbaa3116bac565e009089bb6b36b
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

JanMatCodasip commented:
The following updates were made:

- use explicit runner name (ubuntu-22.04 instead of ubuntu-latest)
- use checkout action v3 (not v2) - silences a Github warning
- attempt to cache dependencies (Spike + toolchain)
- reordered actions so that cached items are handled first
- moved dependencies to /opt/riscv for easier caching
- archive logs from the tests (downloadable artifact)
- more descriptive names for some steps
- changed 'apt' to 'apt-get' to supress a warning

JanMatCodasip commented:
The cache of Spike + RISC-V toolchain reduces the runtime by approx. 40 % (18m 11s --> 11m 13s).

timsifive commented:
> The following updates were made:

Can you put that list in the change description itself?

timsifive commented on .github/workflows/spike-openocd-tests.yml:
```diff
@@ -12,65 +19,99 @@ on: pull_request
 jobs:
   test:
     name: Test debug (Ubuntu)
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-22.04
```
I prefer latest. If we run against latest and it breaks, we can quickly fix it by changing back to the previous version. If the fixed version stops working because github removes it (as just happened for ubuntu-18.04) then the only way to quick fix is to move to a newer version and hope that it works. If it doesn't we have to make the test work with newer ubuntu right now.

JanMatCodasip replied:
I usually prefer to be explicit about requirements/dependencies, that's why I proposed `ubuntu-22.04` (and `ubuntu-latest` is at the moment just an alias to `ubuntu-22.04`). But I do not feel stronly about this one  - I reverted it to `ubuntu-latest`.

timsifive reviewed (dismissed):
Looks good. It's great that the spike build can be cached.
Thanks for these improvements.

JanMatCodasip commented:
> > The following updates were made

> Can you put that list in the change description itself?

Good point - I have squashed the commits and updated the commit message to include the list of changes.

timsifive commented on .github/workflows/spike-openocd-tests.yml:
```diff
@@ -14,63 +21,97 @@ jobs:
 
-      - name: Download Toolchain
-        run: wget --progress=dot:giga https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack/releases/download/v12.2.0-1/xpack-riscv-none-elf-gcc-12.2.0-1-linux-x64.tar.gz
+      - name: Get dependencies from cache
+        id: cache-deps
+        uses: actions/cache@v3
+        with:
+          path: /opt/riscv
+          key: "spike-${{env.SPIKE_COMMIT}}-toolchain-${{env.TOOLCHAIN_URL}}"
```
Can you turn this into 2 caches: one for spike, and one for the toolchain?

The toolchain is quite large but only changes when we update this file. spike changes all the time. It seems better not to store a new copy of the toolchain just because spike changes.

JanMatCodasip replied:
That's a good suggestion - I will do that. (Just a note: It may take me a week or so to get back to this.)

JanMatCodasip committed:
Split cache for spike and toolchain

Change-Id: I423c4ba28e44ec5126786897135fb245e164c664
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

JanMatCodasip committed:
Split cache for spike and toolchain - fix cache keys

Change-Id: I907bdb52f402b17a7829ea1d06cd395518de4cd3
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

JanMatCodasip committed:
Empty commit to re-trigger the CI

Change-Id: I749d44d8f0dde09ce5adf6e2e1ab5a5324f4018f
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

timsifive reviewed (approved):
Looks good. Thanks!
